### DISCUSSION TITLE
[SWCK] Is there a way to config agent separatly rather then in target object, like deployment/statefulset etc..

### DISCUSSION THREAD

**Discussion by Duncan-tree-zhou** [apache/skywalking#9134] [Q&A]
when i use skywalking swck to add agent in my deployment, i had to add dozens of annotations to config agent. And the manual operation should be done in each deployment which i want to add agent to.

**Comment by wu-sheng** [comment]
PIng @hanahmily @dashanji

**Comment by dashanji** [comment]
@Duncan-tree-zhou In the current design, you can change the [default configmap](https://github.com/apache/skywalking-swck/blob/master/docs/java-agent-injector.md#default-configmap) to configure all Deployments/Statefulsets. If you only want to configure some Deployments/Statefulsets with the specific label, you can use the shell like the following.
```sh
#!/usr/bin/env bash
for namespace in `kubectl get ns -lswck-injection=enabled -o name | awk -F '/' '{print $2}'`
do
  for deployment in `kubectl get deployment -l(use your app label here) -n $namespace -o name | awk -F '/' '{print $2}'`
  do
    kubectl -n $namespace patch deployment $deployment -p '{"spec": {"template":{"metadata":{"annotations":{"strategy.skywalking.apache.org/agent.Overlay":"true"}}}} }'
    kubectl -n $namespace patch deployment $deployment -p '{"spec": {"template":{"metadata":{"annotations":{use your agent configuration here}}}} }'
  done
done
```

**Reply by Duncan-tree-zhou** [reply]
@dashanji Thanks, it partially solved my problem. Is it able to specify a new version of javaagent.jar globally?

**Reply by Duncan-tree-zhou** [reply]
i mean by configuring the default configmap, is it able to specify a new version of javaagent.jar or to config the optional plugin globally? running a script is not a revertable operation.

**Reply by dashanji** [reply]
Currently, the default configmap only supports the [agent configuration](https://skywalking.apache.org/docs/skywalking-java/latest/en/setup/service-agent/java-agent/configurations/#table-of-agent-configuration-properties), the version of java agent and the optional belong to [Sidecar](https://github.com/apache/skywalking-swck/blob/master/docs/java-agent-injector.md#configure-sidecar) which can only be configured by `annotation`.

I don't know whether it is a common requirement to set a global configuration for sidecars. If some users need it, I think we can add the feature in the future. BTW, would you like to add it？

**Reply by Duncan-tree-zhou** [reply]
I think upgrading agents is one of the common scenario. Maybe we can refer how opentelemetry-operator does. To define a CRD for instruments, and the operator do the injection depends on the instrument CR. If it's a acceptable approach, i can try to implement this feature.

**Reply by dashanji** [reply]
@Duncan-tree-zhou  Good idea, a CRD looks good for me. It's nice that you would like to add this feature, I will help you with it. If you agree, I will assign an issue to you. What do you think?

**Reply by Duncan-tree-zhou** [reply]
@dashanji i am ok. please assign the issue to me.

**Reply by wu-sheng** [reply]
@Duncan-tree-zhou This is a discussion, not an issue. If you want a full proposal, please submit a [feature] type issue to explain what you are going to do.

**Reply by Duncan-tree-zhou** [reply]
@wu-sheng @dashanji here is the issue.

https://github.com/apache/skywalking/issues/9160

**Reply by dashanji** [reply]
@Duncan-tree-zhou  Thanks for the proposal.
